altos-rust/port/cortex-m0/src/peripheral/usart/control.rs

```diff
@@ -0,0 +1,426 @@
+// necessary values for a specific usart configuration.
+impl UsartCR {
+    pub fn new(base_addr: *const u32) -> Self {
+        UsartCR {
+            cr1: CR1::new(base_addr),
+            cr2: CR2::new(base_addr),
+            cr3: CR3::new(base_addr),
+        }
+    }
+
+    pub fn enable_usart(&self) {
+        self.cr1.enable_usart(true);
+    }
+
+    pub fn disable_usart(&self) {
+        self.cr1.enable_usart(false);
+    }
+
+    pub fn is_usart_enabled(&self) -> bool {
+        self.cr1.is_usart_enabled()
+    }
+
+    pub fn set_word_length(&self, length: WordLength) {
+        self.cr1.set_word_length(length);
+    }
+
+    pub fn set_mode(&self, mode: Mode) {
+        self.cr1.set_mode(mode);
+    }
+
+    pub fn set_parity(&self, parity: Parity) {
+        self.cr1.set_parity(parity);
+    }
+
+    pub fn set_stop_bits(&self, length: Stoplength) {
+        self.cr2.set_stop_bits(length);
+    }
+
+    pub fn enable_over8(&self) {
+        self.cr1.set_over8(true);
+    }
+
+    pub fn disable_over8(&self) {
+        self.cr1.set_over8(false);
+    }
+
+    pub fn get_over8(&self) -> bool {
+        self.cr1.get_over8()
+    }
+
+    pub fn set_hardware_flow_control(&self, hfc: HardwareFlowControl) {
+        self.cr3.set_hardware_flow_control(hfc);
+    }
+}
+
+// ------------------------------------
+/// CR1
+// ------------------------------------
+
+/// Word length can be 7, 8, or 9 bits.
+#[derive(Copy, Clone)]
+pub enum WordLength {
+    Seven,
+    Eight,
+    Nine,
+}
+
+pub enum Mode {
+    Receive,
+    Transmit,
+    All,
+}
+
+pub enum Parity {
+    None,
+    Even,
+    Odd,
+}
+
+#[derive(Copy, Clone)]
+struct CR1 {
+    base_addr: *const u32,
+}
+
+impl Register for CR1 {
+    fn new(base_addr: *const u32) -> Self {
+        CR1 { base_addr: base_addr }
+    }
+
+    fn base_addr(&self) -> *const u32 {
+        self.base_addr
+    }
+
+    fn mem_offset(&self) -> u32 {
+        CR1_OFFSET
+    }
+}
+
+impl CR1 {
+    // Enables and disables USARTx based on bool variable passed in.
+    fn enable_usart(&self, enable: bool) {
+        unsafe {
+            let mut reg = self.addr();
+            if enable {
+                *reg |= CR1_UE;
+            }
+            else {
+                *reg &= !(CR1_UE);
+            }
+        }
+    }
+
+    // Checks if usart is enabled.
+    fn is_usart_enabled(&self) -> bool {
+        unsafe {
+            *self.addr() & CR1_UE != 0
+        }
+    }
+
+    // Sets wordlength to 7, 8, or 9 bits.
+    fn set_word_length(&self, length: WordLength) {
+        let mask = match length {
+            WordLength::Seven => CR1_M1,
+            WordLength::Eight => 0,
+            WordLength::Nine => CR1_M0,
+        };
+
+        unsafe {
+            let mut reg = self.addr();
+            *reg &= !(CR1_M0 | CR1_M1);
+            *reg |= mask;
+        }
+    }
+
+    // Sets mode for receive(Rx), transmit(Tx) or both(RxTx)
+    fn set_mode(&self, mode: Mode) {
```

Comment:
I do not explicitly have the option to have both bits off. It would implicitly happen if the register is created and this function is not called. In my opinion, there is no need to explicitly have the option to turn both bits off as I'm not sure what the point would be. Should I explicitly add this functionality?